jetty9-compat-base/src/main/java/com/google/apphosting/runtime/jetty9/NoOpSessionManager.java

```diff
@@ -0,0 +1,237 @@
+ * is distributed on an "AS-IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+
+package com.google.apphosting.runtime.jetty9;
+
+import java.util.EventListener;
+import java.util.Set;
+
+import javax.servlet.ServletContext;
+import javax.servlet.SessionCookieConfig;
+import javax.servlet.SessionTrackingMode;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpSession;
+
+import org.eclipse.jetty.http.HttpCookie;
+import org.eclipse.jetty.server.Server;
+import org.eclipse.jetty.server.SessionIdManager;
+import org.eclipse.jetty.server.SessionManager;
+import org.eclipse.jetty.server.handler.ContextHandler;
+import org.eclipse.jetty.server.session.SessionHandler;
+import org.eclipse.jetty.util.component.ContainerLifeCycle;
+
+/**
+ * Does the bare minimum to create a NoOpSession. This session manager should be used only when
+ * app-cfg.xml has disabled sessions. The purpose of it is to provide an implementation with just
+ * enough compliance with the servlet session api so that frameworks that require the session api
+ * will still work, albeit without the costs associated with distributed persistent sessions (ie
+ * sessions enabled in app-cfg.xml).
+ */
+public class NoOpSessionManager extends ContainerLifeCycle implements SessionManager {
+
+  private SessionIdManager _idManager;
+  private SessionHandler _handler;
+  private ServletContext _context;
+
+  /**
+   *
+   */
+  public NoOpSessionManager() {
+  }
+
+
+  @Override
+  protected void doStart() throws Exception {
+    _context = ContextHandler.getCurrentContext();
+    final Server server = _handler.getServer();
+
+    synchronized (server) {
+      if (_idManager == null) {
+        _idManager = server.getSessionIdManager();
+        if (_idManager == null) {
+          // Create the NoOpSessionIdManager and set it as the shared
+          // SessionIdManager for the Server, being careful NOT to use
+          // the webapp context's classloader, otherwise if the context
+          // is restarted, the classloader is leaked.
+          ClassLoader oldLoader = Thread.currentThread().getContextClassLoader();
+          ClassLoader serverLoader = server.getClass().getClassLoader();
+          try {
+            Thread.currentThread().setContextClassLoader(serverLoader);
+            _idManager = new NoOpSessionIdManager();
+            server.setSessionIdManager(_idManager);
+            server.manage(_idManager);
+            _idManager.start();
+          } finally {
+            Thread.currentThread().setContextClassLoader(oldLoader);
+          }
+        }
+
+        // server session id is never managed by this manager
+        addBean(_idManager, false);
+      }
+    }
+    super.doStart();
+  }
+
+
+  @Override
+  public HttpSession getHttpSession(String id) {
+    return null;
```

Comment:
This is called only when a request enters a new context. Thus if a forward is done to the same context, it isn't called. AFAIK forwards to a different context aren't supported in compat because there is only ever 1 context deployed.